### PR TITLE
Feature/more v3 dataset apis

### PR DESCRIPTION
**Pull request by anwar6953**
Expose the following 3 APIs in v3 handlers.
Currently ignores namespace passed in (will utilize that once datasets are namespaced).

`/v3/namespaces/<namespaceId>/datasets`
`/v3/namespaces/<namespaceId>/datasets/<datasetId>`
`/v3/namespaces/<namespaceId>/datasets<datasetId>/flows`
`/v3/namespaces/<namespaceId>/apps/<appId>/datasets`

https://issues.cask.co/browse/CDAP-1375
Builds:
just this PR: http://builds.cask.co/browse/CDAP-DUT842-5

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java [hunk 1]
```diff
@@ -97,4 +97,52 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
...
+  @GET
+  @Path("/datasets")
+  public void getDatasets(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId) {
+    //TODO: use namespace passed in, once datasets are namespaced (https://issues.cask.co/browse/CDAP-775)
+    dataList(request, responder, store, dsFramework, Data.DATASET, Constants.DEFAULT_NAMESPACE, null, null);
+  }
+
+  /**
+   * Returns a dataset associated with account.
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: namespace

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java [hunk 2]
```diff
@@ -97,4 +97,52 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
     programListByDataAccess(request, responder, store, dsFramework, ProgramType.FLOW, Data.STREAM,
                             namespaceId, streamId);
   }
+
+  /**
+   * Returns a list of dataset associated with account.
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: namespace

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 3]
```diff
@@ -584,14 +584,8 @@ public void changeFlowletStreamConnection(HttpRequest request, HttpResponder res
   @Path("/apps/{app-id}")
   public BodyConsumer deploy(HttpRequest request, HttpResponder responder, @PathParam("app-id") final String appId,
                              @HeaderParam(ARCHIVE_NAME_HEADER) final String archiveName) {
-    try {
       return appLifecycleHttpHandler.deploy(rewriteRequest(request), responder, Constants.DEFAULT_NAMESPACE, appId,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: ``RestMigrationUtils.rewrite..``. But why this change? Also, indentation.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 4]
```diff
@@ -602,13 +596,8 @@ public BodyConsumer deploy(HttpRequest request, HttpResponder responder, @PathPa
...
                              @HeaderParam(ARCHIVE_NAME_HEADER) final String archiveName) {
     // null means use name provided by app spec
-    try {
-      return appLifecycleHttpHandler.deploy(rewriteRequest(request), responder, Constants.DEFAULT_NAMESPACE, null,
-                                            archiveName);
-    } catch (Exception ex) {
-      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Deploy failed: " + ex.getMessage());
-      return null;
-    }
+    return appLifecycleHttpHandler.deploy(rewriteRequest(request), responder, Constants.DEFAULT_NAMESPACE, null,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 5]
```diff
@@ -994,7 +986,7 @@ public void getDatasetsByApp(HttpRequest request, HttpResponder responder,
   @Path("/streams/{stream-id}/flows")
   public void getFlowsByStream(HttpRequest request, HttpResponder responder,
                                @PathParam("stream-id") final String streamId) {
-    appFabricStreamHttpHandler.getFlowsByStream(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
+    appFabricDataHttpHandler.getFlowsByStream(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: indentation

**Comment by bdmogal** [comment]
Couple of minor comments, then LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 3 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: the method it delegates to already catches exception and returns null. This method will never receive any exception. it should simply delegate to v3 handler.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 4 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: same reason as:
https://github.com/caskdata/cdap/pull/1355#discussion_r24970465

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: Makes sense.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 6]
```diff
@@ -584,14 +584,8 @@ public void changeFlowletStreamConnection(HttpRequest request, HttpResponder res
   @Path("/apps/{app-id}")
   public BodyConsumer deploy(HttpRequest request, HttpResponder responder, @PathParam("app-id") final String appId,
                              @HeaderParam(ARCHIVE_NAME_HEADER) final String archiveName) {
-    try {
-      return appLifecycleHttpHandler.deploy(rewriteRequest(request), responder, Constants.DEFAULT_NAMESPACE, appId,
+    return appLifecycleHttpHandler.deploy(rewriteRequest(request), responder, Constants.DEFAULT_NAMESPACE, appId,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: Use RestMigrationUtils before merge. From offline discussion, happening in another PR.

**Comment by bdmogal** [comment]
LGTM pending build.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 6 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: yes, in https://github.com/caskdata/cdap/pull/1348

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java [hunk 7]
```diff
@@ -96,4 +95,48 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
...
+                               @PathParam("namespace-id") String namespaceId,
+                               @PathParam("app-id") final String appId) {
+    dataList(request, responder, store, dsFramework, Data.DATASET, namespaceId, null, appId);
+  }
+
+  /**
+   * Returns all flows associated with a dataset.
+   */
+  @GET
+  @Path("/datasets/{dataset-id}/flows")
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: Per https://issues.cask.co/browse/CDAP-1461, is it possible to add other program types like ``/datasets/{dataset-id}/services`` and ``/datasets/{dataset-id}/mapreduce`` easily?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 6 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: ``RestMigrationUtils``?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: Please change this to ``RESTMigrationUtils``

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 8]
```diff
@@ -972,7 +960,7 @@ public void getWorkersByApp(HttpRequest request, HttpResponder responder, @PathP
   @GET
   @Path("/streams")
   public void getStreams(HttpRequest request, HttpResponder responder) {
-    appFabricStreamHttpHandler.getStreams(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
+    appFabricDataHttpHandler.getStreams(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
                                           Constants.DEFAULT_NAMESPACE);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: indentation

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java [hunk 9]
```diff
@@ -993,7 +981,7 @@ public void getStreamSpecification(HttpRequest request, HttpResponder responder,
   @Path("/apps/{app-id}/streams")
   public void getStreamsByApp(HttpRequest request, HttpResponder responder,
                               @PathParam("app-id") final String appId) {
-    appFabricStreamHttpHandler.getStreamsByApp(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
+    appFabricDataHttpHandler.getStreamsByApp(RESTMigrationUtils.rewriteV2RequestToV3(request), responder,
                                                Constants.DEFAULT_NAMESPACE, appId);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: indentation

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java (hunk 7 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: ah yeah - should be doable.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 6 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: different PR: https://github.com/caskdata/cdap/pull/1348/files#diff-887698b13a9e83c3c873f29bf9633f9fR611

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java (hunk 7 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: cool. all the program types you can then. also add them to the v3 APIs page if you add new APIs.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java (hunk 4 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricHttpHandler.java]: https://github.com/caskdata/cdap/pull/1348/files#diff-887698b13a9e83c3c873f29bf9633f9fR611

**Comment by bdmogal** [comment]
some comments. please address.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java (hunk 7 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: turns out theres a problem with how MapReduceSpec is... datasets aren't added to the MR spec depending on how it is used in the MR:
`useDatasets(X)` vs `@UseDataset(X)`

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: https://issues.cask.co/browse/CDAP-1568

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java [hunk 10]
```diff
@@ -96,4 +95,48 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
...
+  }
+
+  /**
+   * Returns a dataset associated with namespace.
+   */
+  @GET
+  @Path("/datasets/{dataset-id}")
+  public void getDatasetSpecification(HttpRequest request, HttpResponder responder,
+                                      @PathParam("namespace-id") String namespaceId,
+                                      @PathParam("dataset-id") final String datasetId) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: no final

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java [hunk 11]
```diff
@@ -96,4 +95,48 @@ public void getFlowsByStream(HttpRequest request, HttpResponder responder,
...
+  }
+
+  /**
+   * Returns a list of dataset associated with application.
+   */
+  @GET
+  @Path("/apps/{app-id}/datasets")
+  public void getDatasetsByApp(HttpRequest request, HttpResponder responder,
+                               @PathParam("namespace-id") String namespaceId,
+                               @PathParam("app-id") final String appId) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppFabricDataHttpHandler.java]: no final

**Comment by bdmogal** [comment]
two minor comments, LGTM otherwise.
